aggregator_core/src/taskprov.rs

```diff
@@ -0,0 +1,403 @@
+        self.collector_auth_tokens.iter().next_back().unwrap()
+    }
+
+    /// Checks if the given collector authentication token is valid (i.e. matches with an
+    /// authentication token recognized by this task).
+    pub fn check_collector_auth_token(&self, auth_token: &AuthenticationToken) -> bool {
+        self.collector_auth_tokens
+            .iter()
+            .rev()
+            .any(|t| t == auth_token)
+    }
+
+    /// Computes the VDAF verify key using the method defined in [draft-wang-ppm-dap-taskprov][1].
+    ///
+    /// [1]: https://www.ietf.org/archive/id/draft-wang-ppm-dap-taskprov-04.html#name-deriving-the-vdaf-verificat
+    pub fn derive_vdaf_verify_key(
+        &self,
+        task_id: &TaskId,
+        vdaf_instance: &VdafInstance,
+    ) -> SecretBytes {
+        let prk = SALT.extract(self.verify_key_init.as_ref());
+        let info = [task_id.as_ref().as_slice()];
+
+        // Unwrap safety: this function only errors if the OKM length is too long
+        // (<= 255 * HashLength). It is not expected that a VDAF's verify key length will ever
+        // be _that_ long.
+        let length = vdaf_instance.verify_key_length();
+        let okm = prk.expand(&info, VdafVerifyKeyLength(length)).unwrap();
+
+        let mut vdaf_verify_key = vec![0; length];
+        // Same unwrap rationale as above.
+        okm.fill(&mut vdaf_verify_key).unwrap();
+        SecretBytes::new(vdaf_verify_key)
+    }
+}
+
+/// Helper type for using `ring::Prk::expand()`.
+struct VdafVerifyKeyLength(usize);
+
+impl KeyType for VdafVerifyKeyLength {
+    fn len(&self) -> usize {
+        self.0
+    }
+}
+
+/// Newtype for [`task::Task`], which omits certain fields that aren't required for taskprov tasks.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Task(task::Task);
+
+impl Task {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        task_id: TaskId,
+        aggregator_endpoints: Vec<Url>,
+        query_type: QueryType,
+        vdaf: VdafInstance,
+        role: Role,
+        vdaf_verify_keys: Vec<SecretBytes>,
+        max_batch_query_count: u64,
+        task_expiration: Option<Time>,
+        report_expiry_age: Option<Duration>,
+        min_batch_size: u64,
+        time_precision: Duration,
+        tolerable_clock_skew: Duration,
+    ) -> Result<Self, Error> {
+        let task = Self(task::Task::new_without_validation(
+            task_id,
+            aggregator_endpoints,
+            query_type,
+            vdaf,
+            role,
+            vdaf_verify_keys,
+            max_batch_query_count,
+            task_expiration,
+            report_expiry_age,
+            min_batch_size,
+            time_precision,
+            tolerable_clock_skew,
+            None,
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+        ));
+        task.validate()?;
+        Ok(task)
+    }
+
+    fn validate(&self) -> Result<(), Error> {
```

Comment:
Clarification: two implementations of `validate()` or the entire struct?

For `validate()`, this is confounded by the fact that taskprov validation is not really a subset of normal task validation. I could possibly factor out the common parts into 3 functions `validate_normal` `validate_common` `validate_taskprov`, but not sure if that's really an improvement.

Reply:
I buy that validation logic differs across the two kinds of task, but it appears a lot of logic is duplicated across the two. For instance, we will soon change `Task` so that instead of  `aggregator_endpoints: Vec<Url>`, it'll just have two fields `leader_endpoint: Url, helper_endpoint: Url`. When we make that change, I only want to have to update a single `validate()` implementation.